core/src/main/java/com/sequenceiq/cloudbreak/service/image/catalog/VersionBasedImageCatalogService.java

```diff
@@ -57,39 +60,45 @@ public void validate(CloudbreakImageCatalogV3 imageCatalogV3) throws CloudbreakI
 
     @Override
     public ImageFilterResult getImageFilterResult(CloudbreakImageCatalogV3 imageCatalogV3) {
-        Set<String> imageIds = getImageIds(imageCatalogV3.getVersions());
+        Set<String> imageIds = getImageIds(imageCatalogV3);
         LOGGER.debug("{} image id(s) found for Cloudbreak version: {}", imageIds.size(), cbVersion);
         String message = String.format("%d image id(s) found for Cloudbreak version: %s", imageIds.size(), cbVersion);
-        List<Image> images = imageCatalogV3.getImages().getCdhImages().stream()
-                .filter(image -> imageIds.contains(image.getUuid()))
-                .collect(Collectors.toList());
-        return new ImageFilterResult(new Images(null, images, null), message);
+        List<Image> freeipaImages = imageCatalogV3.getImages().getFreeIpaImages();
+        if (!freeipaImages.isEmpty()) {
+            List<Image> images = freeipaImages.stream().filter(image -> imageIds.contains(image.getUuid())).collect(Collectors.toList());
+            return new ImageFilterResult(new Images(null, null, images, null), message);
+        } else {
+            List<Image> cdhImages = imageCatalogV3.getImages().getCdhImages();
+            List<Image> images = cdhImages.stream().filter(image -> imageIds.contains(image.getUuid())).collect(Collectors.toList());
+            return new ImageFilterResult(new Images(null, images, null, null), message);
+        }
     }
 
     private void validateVersion(CloudbreakImageCatalogV3 imageCatalogV3) throws CloudbreakImageCatalogException {
-        if (imageCatalogV3.getVersions() == null || imageCatalogV3.getVersions().getCloudbreakVersions().isEmpty()) {
+        if (imageCatalogV3.getVersions() == null || cloudbreakVersionListProvider.getVersions(imageCatalogV3).isEmpty()) {
             throw new CloudbreakImageCatalogException("Cloudbreak versions cannot be NULL");
         }
     }
 
     private void validateUuids(CloudbreakImageCatalogV3 imageCatalogV3) throws CloudbreakImageCatalogException {
         Stream<String> baseUuids = imageCatalogV3.getImages().getBaseImages().stream().map(Image::getUuid);
         Stream<String> cdhUuids = imageCatalogV3.getImages().getCdhImages().stream().map(Image::getUuid);
-        Stream<String> uuidStream = Stream.of(baseUuids, cdhUuids).
-                reduce(Stream::concat).
-                orElseGet(Stream::empty);
+        Stream<String> freeipaUuids = imageCatalogV3.getImages().getFreeIpaImages().stream().map(Image::getUuid);
+        Stream<String> uuidStream = freeipaUuids.findAny().isPresent() ?
```

Comment:
This fails in case of freeipa as freeipaUuids stream get processed double times. Please fix it and cover the case with unit test to make sure it works for freeipa as well.

Reply:
This issue is solved in such a way that the three image lists were given together, because or only the cdh or the freeipa part is filled. Furthermore more unit test are implemented for this validation part which is affected.